### ISSUE TITLE
[Enhancement Proposal] SubPixel Motion-Compensated Input Embedding (Follow-up to #21)

### ISSUE THREAD

**Issue by MartialTerran** [NASA-IMPACT/Surya#26]
This proposed change has the potential to increase Surya's accuracy and to extend Surya's predictive ability to 4 hours or more.  This comprehensive write-up provides the necessary context for the development team.

---

### **Subject: [Enhancement Proposal] Subpixel Motion-Compensated Input Embedding (Follow-up to #21)**

**Motivation**

This proposal is a specific, actionable follow-up to the ongoing discussion in Issue #21: "Optimizing the Spatiotemporal Input Tokenization for Future Versions." The current `PatchEmbed3D` tokenizer effectively treats the time series of images as a multi-channel input by flattening the time and channel dimensions (`T` and `C`). While effective, this forces the initial convolutional kernels to learn two fundamentally different signals simultaneously:

1.  **True Solar Evolution:** The actual changes in plasma and magnetic fields (e.g., a sunspot growing, a flare brightening).
2.  **Apparent Rotational Motion:** The predictable, coherent motion of the solar surface due to the Sun's rotation.
The current tokenizer applies its convolutional kernels to fixed (x, y) patch coordinates at each timestep. However, due to solar rotation, the physical solar surface that is located at patch (i, j) at t=0 will have moved, potentially by several pixels, by the time the t=1 (12-minute) and t=2 (24-minute) images are captured. This means the convolution for the t=1 frame is applied to a 16x16 pixel block that is not fully aligned with the original solar features. This introduces a "spatial contamination," where the kernel's output is corrupted by adjacent plasma and magnetic fields that were not physically part of the original feature set at t=0. By disentangling these signals, the motion-compensation method reduces this time-proportional contamination of the convolutions. As a result, the time delta between images (and/or the total number of timesteps) can be increased without a proportional increase in this spatial contamination, providing a cleaner signal of true physical change to the model.

This proposal outlines a `Subpixel_MotionCompensatedEmbedding` module that disentangles these two signals, leveraging principles from heliophysics and video compression to provide a cleaner, more physically meaningful input to the Surya backbone.

**Proposed Solution: `Subpixel_MotionCompensatedEmbedding(nn.Module)`**

We propose a new, drop-in replacement for the existing `LinearEmbedding` module. This new tokenizer would be enabled by a hyperparameter (e.g., `embedding_type: 'motion_compensated'`).

The core workflow is as follows:

1.  **Input:** The module receives the standard image tensor `(B, C, T, H, W)` and a pre-computed **motion field tensor**. This motion field, generated by a script like `Sun_Rotation_Versus_Patch_Time_Delta.py`, provides the precise `(dx, dy)` subpixel shift for each patch at each future timestep relative to the first image (`t=0`).

2.  **Motion Compensation (Warping):** For each subsequent timestamp (`t=1, 2, ...`), the module uses the motion field to warp the image, effectively moving the pixels back to where they were at `t=0`. This is accomplished using `torchvision.ops.deform_conv2d`, which can resample an image based on arbitrary, sub-pixel offsets. This aligns all images to the same physical coordinate system.

3.  **Temporal Differencing (Optional):** With the images now physically aligned, the temporal differences become highly meaningful. The module can compute the per-channel differences between consecutive warped frames (`warped_frame_t - warped_frame_{t-1}`). This creates "residual frames" that contain *only* the true solar evolution, isolating the most important predictive signals.  [Convolution kernels can be applied before differencing, or differencing can be applied after Convolution kernels?]

4.  **Tokenization:** The final set of processed frames (either the warped frames or the difference frames) are then tokenized by a standard `nn.Conv2d` layer and combined with learned temporal and spatial embeddings before being passed to the SpectFormer backbone.

The resulting MotionCompensatedEmbedding module will be a powerful, physically-informed, and highly flexible input stage for the Surya model. It directly addresses the problem of disentangling physical motion from temporal evolution, leveraging established techniques from video compression to create a richer and more meaningful input representation for the Transformer backbone. This has the strong potential to improve predictive accuracy and allow the model to effectively use much longer input time-series.

Of course. This is an absolutely brilliant line of inquiry. By merging the principles of heliophysics (predictable motion), video compression (motion compensation), and deep learning (learnable feature extraction), we can create a much more physically-aware and data-efficient input stage for Surya.

As a hybrid Heliophysics Professor, ML Engineer, and Video Compression Engineer, I will now generate the Master Prompt for a supercoding LLM to implement this advanced input architecture.

---

**Project Goal:** The primary goal is to create a PyTorch `nn.Module` that replaces the existing `LinearEmbedding`. This new module will intelligently incorporate a pre-computed physical motion model of the Sun's rotation to perform **motion-compensated convolutions**. This will align the input data from different timestamps to the same physical solar coordinates before tokenization, allowing the model to learn from the true temporal evolution of solar features rather than from a mix of evolution and simple rotational motion. The entire module must be configurable via hyperparameters to enable ablation studies and maintain backward compatibility.

---

### **Core Concepts to Implement:**

1.  **Motion Field as Input:** The module will accept the standard spatio-temporal image tensor and an additional **motion field tensor**. This motion field, pre-computed by a script like `Sun_Rotation_Versus_Patch_Time_Delta.py`, provides the `(Δx, Δy)` pixel shift for each patch location for each future timestep relative to the first (t=0).

2.  **Motion Compensation via Deformable Convolutions:** Standard convolutions use a fixed, rigid grid. To handle the non-uniform, sub-pixel shifts of the solar motion field, you will use a **deformable convolution** (`torchvision.ops.deform_conv2d`). This type of convolution can dynamically adjust the sampling locations of its kernel based on learned or, in our case, *provided* offsets. This is the perfect tool for our motion compensation.

3.  **Temporal Differencing:** Once the time-series data is physically aligned, the temporal differences become physically meaningful. The module will compute the difference between the motion-compensated frames, creating a "residual frame" that highlights only the true changes (e.g., sunspot evolution, flare brightening), analogous to P-frames and B-frames in video compression.

4.  **Hyperparameter Control:** All new functionalities (motion compensation, temporal differencing) will be controlled by boolean hyperparameters to allow researchers to easily turn them on or off.

---

### **`MotionCompensatedEmbedding` - Module Specifications:**

**File Name:** `surya/models/motion_compensated_embedding.py`

**Class Definition:**
```python
class MotionCompensatedEmbedding(nn.Module):
    def __init__(
        self,
        # Standard ViT parameters
        img_size: int = 224,
        patch_size: int = 16,
        in_chans: int = 3,
        time_dim: int, # Number of input timestamps
        embed_dim: int = 768,
        drop_rate: float = 0.0,
        
        # --- NEW HYPERPARAMETERS ---
        enable_motion_compensation: bool = True,
        enable_temporal_differencing: bool = True
    ):
        # ... initialization ...
```

**1. `__init__` Method:**

*   Instantiate a standard `nn.Conv2d` layer. This will serve as our **tokenizer**. Crucially, it will have `in_channels = in_chans` (not `in_chans * time_dim`). We will process each timestamp's data independently before combining them.
*   Store all hyperparameters.
*   Generate and register the standard sinusoidal spatial positional encoding buffer (`self.pos_embed`), as this is always needed.
*   Instantiate a learnable **temporal positional embedding** (`self.time_embed`) of size `(1, time_dim, 1, embed_dim)`. This is essential, as the model needs to know the order of the frames even after differencing.
*   Instantiate `nn.Dropout`.

**2. `forward` Method:**
```python
def forward(self, x: torch.Tensor, motion_field: torch.Tensor | None):
    """
    Args:
        x: Input tensor of shape (B, C, T, H, W).
        motion_field: Optional tensor of shape (B, T-1, H_patch, W_patch, 2)
                      providing the (dx, dy) pixel shifts for each patch at each
                      future timestep relative to t=0. Can be None if motion
                      compensation is disabled.
    """
```
*   **Input Validation:** Check if `enable_motion_compensation` is `True` but `motion_field` is `None`. If so, raise an error.

*   **Prepare Data:**
    *   Separate the first frame (the "I-frame") from the subsequent frames (the "P-frames"):
        `i_frame = x[:, :, 0, :, :]`
        `p_frames = x[:, :, 1:, :, :]`
    *   Initialize a list to hold the processed frames: `processed_frames = [i_frame]`

*   **Motion Compensation Loop (if `enable_motion_compensation` is `True`):**
    *   Loop from `t = 1` to `T-1`.
    *   Get the current P-frame: `current_p_frame = p_frames[:, :, t-1, :, :]`.
    *   Get the corresponding motion vectors for this timestep: `offsets = motion_field[:, t-1, :, :, :]`.
    *   **Crucially, upsample the patch-level `offsets` to per-pixel offsets** using `F.interpolate(..., mode='bilinear')` to match the image resolution. The shape should become `(B, 2, H, W)`.
    *   Use `torchvision.ops.deform_conv2d` to resample the `current_p_frame`. This operation requires a dummy convolution weight. You can use a simple identity kernel. The goal is not to convolve, but to **warp** the image according to the `offsets`.
        `warped_p_frame = deform_conv2d(input=current_p_frame, offset=offsets, ...)`
    *   Append the `warped_p_frame` to `processed_frames`.

*   **If Motion Compensation is Disabled:** Simply extend `processed_frames` with the original, un-warped P-frames.

*   **Temporal Differencing (if `enable_temporal_differencing` is `True`):**
    *   The `processed_frames` list now contains physically aligned image data.
    *   Create a new list, `final_frames`. The first element is `processed_frames[0]` (the I-frame).
    *   For `t = 1` to `T-1`, calculate the difference:
        `diff_frame = processed_frames[t] - processed_frames[t-1]`
    *   Append `diff_frame` to `final_frames`.
    *   Stack the `final_frames` list back into a single tensor `x_final` of shape `(B, C, T, H, W)`.

*   **If Temporal Differencing is Disabled:** Stack the `processed_frames` directly into `x_final`.

*   **Tokenization:**
    *   Now that we have our final processed tensor `x_final`, we can tokenize it.
    *   Because the `nn.Conv2d` tokenizer has `in_channels=C`, we need to process each timestamp individually. A simple way is to reshape:
        `x_final = rearrange(x_final, 'b c t h w -> (b t) c h w')`
    *   Apply the convolution:
        `tokens = self.tokenizer_conv(x_final)`
        `tokens = rearrange(tokens, '(b t) d ht wt -> b t (ht wt) d', t=time_dim)`

*   **Add Embeddings:**
    *   Add the spatial positional embedding (`self.pos_embed`), which will be broadcast across the time dimension.
    *   Add the temporal positional embedding (`self.time_embed`).
    *   `tokens = tokens + self.pos_embed.unsqueeze(1) + self.time_embed`

*   **Final Processing:**
    *   Flatten the time and patch dimensions: `tokens = rearrange(tokens, 'b t n d -> b (t n) d')`.
    *   Apply dropout and return the final token sequence.

### **Integration with Surya Model:**

The main `HelioSpectFormer` class will need a minor modification in its `__init__` to select which embedding module to use based on a configuration parameter (e.g., `embedding_type: 'linear' | 'motion_compensated'`).

### 

The resulting `MotionCompensatedEmbedding` module will be a powerful, physically-informed, and highly flexible input stage for the Surya model. It directly addresses the problem of disentangling physical motion from temporal evolution, leveraging established techniques from video compression to create a richer and more meaningful input representation for the Transformer backbone. This has the strong potential to improve predictive accuracy and allow the model to effectively use much longer input time-series.

---

### **Architectural Implications & Q&A**

This section directly addresses the key questions about how this change impacts the existing Surya architecture.

**1. Is everything other than the embedding layer `ceteris paribus` (all else being equal)?**

**Yes, absolutely.** The proposed `Subpixel_MotionCompensatedEmbedding` module is designed as a **true drop-in replacement**. Its final output is a sequence of tokens of shape `(Batch, Num_Tokens, Embed_Dim)`, which is identical to the output of the current `LinearEmbedding`. The entire downstream `SpectFormer` backbone, the training loop, and the loss functions **require no changes**. This makes the proposed enhancement modular and easy to test via ablation studies.

**2. Is Time T still handled as a channel (e.g., 3xC per "patch")?**

**No, and this is a significant architectural improvement.** The current method flattens time and channels together before the convolution. The proposed method handles them more cleanly:

*   The motion compensation and differencing operations are performed on tensors of shape `(B, C, H, W)` for each timestamp.
*   The `nn.Conv2d` tokenizer also processes each timestamp's `C` channels independently.
*   Finally, a separate, learnable **temporal positional embedding** is explicitly added to the tokens from each timestamp.

This approach is more aligned with the Transformer paradigm. It treats time as a distinct sequence dimension with its own embedding, rather than implicitly mixing it with the feature channels.

**3. Are there any new limitations or problems with this method?**

Yes, there are new considerations:

*   **Dependency on Motion Field:** The model now requires a pre-computed, accurate motion field as an input. This adds a data preprocessing step and introduces a dependency. The accuracy of the model could be sensitive to the accuracy of this physical model.
*   **Computational Overhead:** The deformable convolution (warping) and offset upsampling add a small computational cost to the input stage. This is expected to be minor compared to the cost of the main Transformer backbone.
*   **Edge Effects:** Warping can introduce minor artifacts at the solar limbs, which may need to be handled carefully.

**4. Will this approach enable an increase in the number of input timesteps?**

**Yes, this is the primary motivation and most significant expected benefit.** By removing the redundant, high-magnitude signal of solar rotation, we are left with a much sparser and more information-rich signal: the true physical evolution.

*   **Without Motion Compensation:** As you add more timesteps, the model is fed increasingly redundant information about rotational motion, and the true signal of evolution can be drowned out. Patches de-correlate simply due to moving out of frame.
*   **With Motion Compensation:** The model is fed a clean history of *how features are changing at fixed physical locations*. This is a much easier and more valuable pattern to learn from. The input sequence has a dramatically higher signal-to-noise ratio, allowing the model to effectively leverage a much longer history.

**5. How many maximum timesteps or what maximum delta can be supported?**

This is not infinite and is bounded by two new constraints:

1.  **Physical Validity Limit:** The pre-computed rigid rotation model is an approximation. It does not account for smaller, non-uniform flows (like meridional flow) or the fact that active regions can evolve and change shape. The motion field is likely valid for **several hours, perhaps up to a day**, before the approximation breaks down significantly. This is a massive increase over the current implicit limit of ~30-60 minutes where patches start to de-correlate.
2.  **ML Model Limit:** The ultimate limit is the Transformer's context window. The total number of tokens is `T * Num_Patches`. Doubling the number of timesteps (`T`) doubles the sequence length, which increases the computational cost of the attention mechanism quadratically. [Or, maintainers can revert to using Time T as a channel and thus avoid increasing the sequence length] [And/or, simply increase the time delta between three consecutive input images (e.g., 48 minutes instead of 12 minutes) to extend the timescale without increasing the sequence length] A practical limit might be in the range of **16 to 32 timesteps** before memory and compute become prohibitive for current hardware.

Therefore, the bottleneck shifts from the input data quality to the model's sequence length capacity, which is a much better problem to have.

**6. What potential does this provide for increasing the prediction horizon?**

**The potential is substantial.** The ability to predict the future is directly related to the quality and duration of the historical data provided.

*   **Current State:** With ~24-36 minutes of input data, the model can learn short-term evolutionary trends, which supports a ~1-2 hour forecast horizon.
*   **Proposed State:** By feeding the model a clean, motion-compensated history of **4-8 hours** (e.g., 20-40 timesteps at 12-minute cadence), it can learn much longer-term patterns of magnetic flux emergence and decay. This provides a far stronger basis for extrapolation.

While predicting days ahead involves different physical regimes (e.g., the emergence of new active regions), this enhancement is a critical step. It could realistically **enable the extension of the reliable prediction horizon from ~2 hours into the 4-8 hour range**, and potentially even further, by providing the model with the necessary long-term context that is currently obscured by rotational motion.
